kmk/keys.py

```diff
@@ -21,40 +21,307 @@
+    if candidate in names:
+        return make_mod_key(code=code, names=names, *args, **kwargs)
 
 
-def first_truthy(candidate, *funcs):
-    for func in funcs:
-        result = func(candidate)
-        if result is not None:
-            return result
+def maybe_make_key(candidate, code, names, *args, **kwargs):
+    if candidate in names:
+        return make_key(code=code, names=names, *args, **kwargs)
 
 
-def maybe_make_mod_key(candidate, code, names):
+def maybe_make_shifted_key(candidate, code, names, *args, **kwargs):
     if candidate in names:
-        return make_mod_key(code=code, names=names)
+        return make_shifted_key(code=code, names=names, *args, **kwargs)
 
 
-def maybe_make_key(candidate, code, names):
+def maybe_make_consumer_key(candidate, code, names, *args, **kwargs):
     if candidate in names:
-        return make_key(code=code, names=names)
+        return make_consumer_key(code=code, names=names, *args, **kwargs)
 
 
-def maybe_make_shifted_key(candidate, code, names):
+def maybe_make_argumented_key(
+    candidate,
+    validator=lambda *validator_args, **validator_kwargs: object(),
+    names=tuple(),  # NOQA
+    *constructor_args,
+    **constructor_kwargs,
+):
     if candidate in names:
-        return make_shifted_key(code=code, names=names)
+        return make_argumented_key(
+            validator, names, *constructor_args, **constructor_kwargs
+        )
 
 
-def maybe_make_consumer_key(candidate, code, names):
-    if candidate in names:
-        return make_consumer_key(code=code, names=names)
+def maybe_make_alpha_key(candidate):
+    if len(candidate) == 1:
+        candidate_upper = candidate.upper()
+        if candidate_upper in ALL_ALPHAS:
+            return make_key(
+                code=4 + ALL_ALPHAS.index(candidate_upper),
+                names=(
+                    candidate_upper,
+                    candidate.lower(),
+                ),
+            )
+
+
+def maybe_make_numeric_key(candidate):
+    if candidate in ALL_NUMBERS or candidate in ALL_NUMBER_ALIASES:
+        try:
+            offset = ALL_NUMBERS.index(candidate)
+        except ValueError:
+            offset = ALL_NUMBER_ALIASES.index(candidate)
+
+        return make_key(
+            code=30 + offset, names=(ALL_NUMBERS[offset], ALL_NUMBER_ALIASES[offset])
+        )
+
+
+__unicode_noop = UnicodeModeKeyMeta(UnicodeMode.NOOP)
+__unicode_ibus = UnicodeModeKeyMeta(UnicodeMode.IBUS)
+__unicode_ralt = UnicodeModeKeyMeta(UnicodeMode.RALT)
+__unicode_winc = UnicodeModeKeyMeta(UnicodeMode.WINC)
+
+KEY_GENERATION_FUNCTIONS = (
+    # NO and TRNS are functionally identical in how they (don't) mutate
+    # the state, but are tracked semantically separately, so create
+    # two keys with the exact same functionality
+    (
+        maybe_make_key,
+        (None, ('NO', 'XXXXXXX')),
+        {'on_press': handlers.passthrough, 'on_release': handlers.passthrough},
+    ),
```

Comment:
This is definitely more memory intensive than the lambdas. You're creating a dictionary with string indices for every key.